### ISSUE TITLE
[Bug]: MalformedPolicyDocument not detected at plan step

### ISSUE THREAD

**Issue by nicolaschambrier** [hashicorp/terraform-provider-aws#37900]
### Terraform Core Version

1.5.5

### AWS Provider Version

5.53.0

### Affected Resource(s)

* aws_iam_policy

### Expected Behavior

By mistake we included a `aws_iam_policy` statement with an empty list of resources 

```
data "aws_iam_policy_document" "apps_ci" {
  # [...] a bunch of valid statements
  statement {
    actions = [
      "sqs:*",
    ]
    resources = var.queues_arn
  }
}
```

This failed at apply time, causing the policy to obviously not being applied.

### Actual Behavior

It's a totally predictible error, that should have failed at plan time. It could have caused severe deployment issues due to partial apply.

### Terraform Configuration Files

```tf
data "aws_iam_policy_document" "this" {
  statement {
    actions = [
      "sqs:*",
    ]
    resources = []
  }
}
resource "aws_iam_policy" "lambda_deployment_policies" {
  name        = "policy"
  policy = data.aws_iam_policy_document.this.json
}
```

### Steps to Reproduce

- run `terraform plan` ➡️ no error
- run `terraform apply` ➡️ fail

### Would you like to implement a fix?

None

**Comment by sourcefrog** [comment]
I'm potentially interested in working on this. Would a PR to do this be in principle acceptable? Is there any prior art for turning optional preflight checks on and off?

AWS has a ValidatePolicy API which Terraform could call at plan time.

**Comment by justinretzolk** [comment]
Relates #18935

**Comment by sourcefrog** [comment]
Good point about Access Analyzer potentially being blocked by an SCP or something.

Maybe we could have an option on the resource or even on the provider that opts in to plan-time validation?

**Comment by sidewinder12s** [comment]
Yes that was my thought was make it an optional parameter on the data resource to run the generated IAM Policy Document through Access Analyzer for policy validation only.

That or just add a data resource against IAM Access Analyzer you'd run your policy through, though that is less nice, but that might fit within the Terraform ecosystem more.
